File: FBDeviceControl/Management/FBAMDeviceServiceManager.m
/*
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

#import "FBAMDeviceServiceManager.h"

#import "FBAMDevice.h"
#import "FBDeviceControlError.h"
#import "FBAMDevice+Private.h"

@interface FBAMDeviceServiceManager_HouseArrest : NSObject<FBFutureContextManagerDelegate>

@property (nonatomic, weak, readonly) FBAMDevice *device;
@property (nonatomic, copy, readonly) NSString *bundleID;
@property (nonatomic, assign, readonly) AFCCalls calls;

@end

@implementation FBAMDeviceServiceManager_HouseArrest

@synthesize contextPoolTimeout = _contextPoolTimeout;

- (instancetype)initWithDevice:(FBAMDevice *)device bundleID:(NSString *)bundleID calls:(AFCCalls)calls serviceTimeout:(nullable NSNumber *)serviceTimeout
{
  self = [super init];
  if (!self) {
    return nil;
  }

  _device = device;
  _bundleID = bundleID;
  _calls = calls;
  _contextPoolTimeout = serviceTimeout;

  return self;
}

- (FBFuture<FBAFCConnection *> *)prepare:(id<FBControlCoreLogger>)logger
{
  AFCConnectionRef afcConnection = NULL;
  [logger logFormat:@"Starting house arrest for '%@'", self.bundleID];
  int status = self.device.calls.CreateHouseArrestService(
    self.device.amDeviceRef,
    (__bridge CFStringRef _Nonnull)(self.bundleID),
    NULL,
    &afcConnection
  );
  if (status != 0) {
    NSString *internalMessage = CFBridgingRelease(self.device.calls.CopyErrorText(status));
    return [[[FBDeviceControlError
      describeFormat:@"Failed to start house_arrest service for '%@' with error 0x%x (%@)", self.bundleID, status, internalMessage]
      logger:logger]
      failFuture];
  }
  FBAFCConnection *connection = [[FBAFCConnection alloc] initWithConnection:afcConnection calls:self.calls logger:logger];
  return [FBFuture futureWithResult:connection];
}

- (FBFuture<NSNull *> *)teardown:(FBAFCConnection *)connection logger:(id<FBControlCoreLogger>)logger
{
  [logger logFormat:@"Closing connection to House Arrest for '%@'", self.bundleID];
  NSError *error = nil;
  if (![connection closeWithError:&error]) {
    [logger logFormat:@"Failed to close House Arrest for '%@' with error %@", self.bundleID, error];
    return [FBFuture futureWithError:error];
  } else {
    [logger logFormat:@"Closed House Arrest service for '%@'", self.bundleID];
    return FBFuture.empty;
  }
}

- (NSString *)contextName
{
  return [NSString stringWithFormat:@"house_arrest_%@", self.bundleID];
}

- (BOOL)isContextSharable
{
  return NO;
}

@end

@interface FBAMDeviceServiceManager ()

@property (nonatomic, weak, readonly) FBAMDevice *device;
@property (nonatomic, copy, nullable, readonly) NSNumber *serviceTimeout;
@property (nonatomic, strong, readonly) NSMutableDictionary<NSString *, FBFutureContextManager<FBAFCConnection *> *> *houseArrestManagers;
@property (nonatomic, strong, readonly) NSMutableDictionary<NSString *, FBAMDeviceServiceManager_HouseArrest *> *houseArrestDelegates;

@end

@implementation FBAMDeviceServiceManager

#pragma mark Initializers

+ (instancetype)managerWithAMDevice:(FBAMDevice *)device serviceTimeout:(nullable NSNumber *)serviceTimeout
{
  return [[self alloc] initWithAMDevice:device serviceTimeout:serviceTimeout];
}

- (instancetype)initWithAMDevice:(FBAMDevice *)device serviceTimeout:(nullable NSNumber *)serviceTimeout
{
  self = [super init];
  if (!self) {
    return nil;
  }

  _device = device;
  _serviceTimeout = serviceTimeout;
  _houseArrestManagers = [NSMutableDictionary dictionary];
  _houseArrestDelegates = [NSMutableDictionary dictionary];

  return self;
}

#pragma mark Public Services

- (FBFutureContextManager<FBAFCConnection *> *)houseArrestAFCConnectionForBundleID:(NSString *)bundleID afcCalls:(AFCCalls)afcCalls
{
  FBFutureContextManager<FBAFCConnection *> *manager = self.houseArrestManagers[bundleID];
  if (manager) {
    return manager;
  }
  FBAMDeviceServiceManager_HouseArrest *delegate = [[FBAMDeviceServiceManager_HouseArrest alloc] initWithDevice:self.device bundleID:bundleID calls:afcCalls serviceTimeout:self.serviceTimeout];
  manager = [FBFutureContextManager managerWithQueue:self.device.workQueue delegate:delegate logger:self.device.logger];
  self.houseArrestManagers[bundleID] = manager;
  self.houseArrestDelegates[bundleID] = delegate;
  return manager;
}

@end

File: FBDeviceControl/Management/FBAMDeviceServiceManager.swift
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

@preconcurrency import FBControlCore
import Foundation

/// The ways house-arrest service management can fail, as data rather than assembled strings.
enum FBAMDeviceServiceError: Error {
  case houseArrestStartFailed(bundleID: String, status: Int32, message: String)
  case houseArrestConnectionMissing(bundleID: String)
  case notAnAFCConnection(context: String)
  case secureStartServiceFailed(service: String, status: Int32, message: String)
  case deviceNotConnected(service: String)
  case notAMDeviceBacked(service: String)
}

extension FBAMDeviceServiceError: LocalizedError {
  var errorDescription: String? {
    switch self {
    case let .houseArrestStartFailed(bundleID, status, message):
      return "Failed to start house_arrest service for '\(bundleID)' with error 0x\(String(status, radix: 16)) (\(message))"
    case let .houseArrestConnectionMissing(bundleID):
      return "No house_arrest connection was returned for '\(bundleID)'"
    case let .notAnAFCConnection(context):
      return "\(context) is not an FBAFCConnection"
    case let .secureStartServiceFailed(service, status, message):
      return "SecureStartService of \(service) Failed with 0x\(String(status, radix: 16)) \(message)"
    case let .deviceNotConnected(service):
      return "Cannot start service \(service): device is not connected"
    case let .notAMDeviceBacked(service):
      return "Cannot start service \(service): the device is not AMDevice backed"
    }
  }
}

/// The house arrest service for a single bundle id, and the AFC connection it vends.
///
/// One consumer uses the connection at a time and the rest wait for it. When the last consumer
/// releases it the connection is not closed but pooled for `reuseTimeout`, so a following
/// operation on the same bundle re-uses it instead of paying for another
/// `AMDeviceCreateHouseArrestService`. With no timeout it is closed as soon as it is released.
///
/// `@unchecked Sendable`: all mutable state is guarded by `lock`.
final class FBHouseArrestService: @unchecked Sendable {

  // MARK: - Properties

  private weak var device: FBAMDevice?
  private let bundleID: String
  private let afcCalls: AFCCalls
  private let reuseTimeout: TimeInterval?
  private let logger: any FBControlCoreLogger

  private let lock = NSLock()
  private var connection: FBAFCConnection?
  private var inUse = false
  private var waiters: [CheckedContinuation<Void, Never>] = []
  private var idleTeardown: Task<Void, Never>?

  // MARK: - Initializers

  init(device: FBAMDevice, bundleID: String, afcCalls: AFCCalls, reuseTimeout: TimeInterval?) {
    self.device = device
    self.bundleID = bundleID
    self.afcCalls = afcCalls
    self.reuseTimeout = reuseTimeout
    self.logger = device.logger.withName("house_arrest_\(bundleID)")
  }

  // MARK: - Public Methods

  /// Waits for exclusive use of the connection, opening it if it is not already established.
  ///
  /// Every call that returns a connection must be paired with a `release()`.
  func acquire() async throws -> FBAFCConnection {
    await takeExclusiveUse()
    if let established = establishedConnection() {
      logger.log("Re-using the existing house arrest connection for '\(bundleID)'")
      return established
    }
    do {
      let opened = try openConnection()
      store(opened)
      return opened
    } catch {
      // Use was taken before the connection could be opened, so it has to be handed on from here.
      release()
      throw error
    }
  }

  /// Gives up use of the connection, passing it to the next waiter or pooling it for reuse.
  func release() {
    lock.lock()
    // Waiters are served back to front. `inUse` stays set: use passes straight on to the consumer
    // resumed here.
    if let next = waiters.popLast() {
      lock.unlock()
      next.resume()
      return
    }
    inUse = false
    idleTeardown?.cancel()
    idleTeardown = nil
    guard let reuseTimeout else {
      let closing = connection
      connection = nil
      lock.unlock()
      close(closing)
      return
    }
    idleTeardown = Task { [weak self] in
      try? await Task.sleep(nanoseconds: UInt64(reuseTimeout * Double(NSEC_PER_SEC)))
      guard !Task.isCancelled else {
        return
      }
      self?.closeIfUnused()
    }
    lock.unlock()
    logger.log("No more consumers of house arrest for '\(bundleID)', pooling it for \(reuseTimeout) seconds of inactivity")
  }

  // MARK: - Private

  private func takeExclusiveUse() async {
    await withCheckedContinuation { continuation in
      lock.lock()
      idleTeardown?.cancel()
      idleTeardown = nil
      if inUse {
        waiters.append(continuation)
        lock.unlock()
        return
      }
      inUse = true
      lock.unlock()
      continuation.resume()
    }
  }

  private func establishedConnection() -> FBAFCConnection? {
    lock.lock()
    defer { lock.unlock() }
    return connection
  }

  private func store(_ connection: FBAFCConnection) {
    lock.lock()
    self.connection = connection
    lock.unlock()
  }

  private func openConnection() throws -> FBAFCConnection {
    guard let device else {
      throw FBDeviceNilError.deviceNil
    }
    logger.log("Starting house arrest for '\(bundleID)'")
    var afcConnection: Unmanaged<AnyObject>?
    let status =
      device.calls.CreateHouseArrestService?(
        device.amDeviceRef,
        bundleID as CFString,
        nil,
        &afcConnection
      ) ?? -1
    guard status == 0 else {
      let message = device.calls.CopyErrorText?(status)?.takeRetainedValue() as String? ?? "unknown"
      throw FBAMDeviceServiceError.houseArrestStartFailed(bundleID: bundleID, status: status, message: message)
    }
    guard let afcConnection else {
      throw FBAMDeviceServiceError.houseArrestConnectionMissing(bundleID: bundleID)
    }
    return FBAFCConnection(connection: afcConnection.takeUnretainedValue(), calls: afcCalls, logger: logger)
  }

  /// The idle window is best-effort cancelled rather than reliably so, hence the second check for
  /// a consumer that took the connection while it was running out.
  private func closeIfUnused() {
    lock.lock()
    if inUse {
      lock.unlock()
      logger.log("Not tearing down house arrest for '\(bundleID)' as it has a consumer again")
      return
    }
    let closing = connection
    connection = nil
    lock.unlock()
    close(closing)
  }

  private func close(_ connection: FBAFCConnection?) {
    guard let connection else {
      return
    }
    logger.log("Closing the connection to house arrest for '\(bundleID)'")
    do {
      try connection.close()
      logger.log("Closed house arrest for '\(bundleID)'")
    } catch {
      logger.log("Failed to close house arrest for '\(bundleID)' with error \(error)")
    }
  }
}

/// The house arrest services a device has opened, one per bundle id.
///
/// `@unchecked Sendable`: the service map is guarded by `lock`.
final class FBAMDeviceServiceManager: @unchecked Sendable {

  private weak var device: FBAMDevice?
  private let serviceTimeout: TimeInterval?
  private let lock = NSLock()
  private var houseArrestServices: [String: FBHouseArrestService] = [:]

  // MARK: Initializers

  init(device: FBAMDevice, serviceTimeout: TimeInterval?) {
    self.device = device
    self.serviceTimeout = serviceTimeout
  }

  // MARK: Public Services

  /// The house arrest service for `bundleID`, created on first use and kept for the device's life.
  func houseArrestService(forBundleID bundleID: String, afcCalls: AFCCalls) -> FBHouseArrestService {
    lock.lock()
    defer { lock.unlock() }
    if let existing = houseArrestServices[bundleID] {
      return existing
    }
    guard let device else {
      preconditionFailure("Device is nil when creating house arrest connection for '\(bundleID)'")
    }
    let service = FBHouseArrestService(device: device, bundleID: bundleID, afcCalls: afcCalls, reuseTimeout: serviceTimeout)
    houseArrestServices[bundleID] = service
    return service
  }
}
